### DISCUSSION TITLE
Find open handles of a program

### DISCUSSION THREAD

**Discussion by stdedos** [winsiderss/systeminformer#819] [Questions / Answers] [answered]
I have seen the very helpful feature of PH to help track down open file handles of e.g. a program keeping a lock on a file, and Windows not be that smart about telling who's responsible. However, ... that doesn't always work.

I hate to ask to be spoonfed... but here it goes:
![image](https://user-images.githubusercontent.com/133706/109727355-e28ea780-7bbc-11eb-9094-6b096fff9b72.png)

I don't see anything holding an open handle on the D: drive.
Of course, maybe, someone can claim that the open handle is on the device itself (which I don't know how to look for) rather than "a plain file with a plain path".

Any feedback welcome!

**Comment by Biswa96** [comment]
PH is working correctly. Here is an example screenshot with Microsoft Windows File Explorer.

![Capture](https://user-images.githubusercontent.com/31443074/109752372-50ab8c80-7c06-11eb-8f68-e03488ef1c39.PNG)

Maybe that program is not working the way you are thinking it would be.

**Reply by stdedos** [reply]
Yeah, of course - I am not saying that this program is 100% doing what it's supposed to do correctly verifiably.
With the last build https://www.ltr-data.se/opencode.html/#ImDisk at 2018, all bets could be off.

**Comment by dmex** [comment]
Finding handles or dlls requires access to the original process to enumerate its handles. So to be able to find all handles you need to:

1) Enable the kernel driver. You'll see a "+" symbol in the window title when the driver is connected.
2) Running as administrator. Required to search handles from the System process and other elevated processes.

If you don't have these enabled then you won't be able to locate/search for handles properly... Also more recently some 'security' products such as Avast/AVG are currently using DKOM (direct kernel object manipulation) exploitation for zeroing the names of handles so you cannot find open handles to disk/usb devices created by their software.

**Reply by stdedos** [reply]
I think I am at that level
![image](https://user-images.githubusercontent.com/133706/109924377-dee84700-7cc8-11eb-9ae2-bdc4c36faf89.png)

And I have gotten rid of Avast! on my new rig 😀

**Comment by diversenok** [comment, marked as answer]
Have you tried searching for `\Device\ImDisk0` instead of a drive letter? I remember [some compatibility issues](https://wj32.org/processhacker/forums/viewtopic.php?f=33&t=3419&p=10598#p10598) were preventing Process Hacker from correctly converting native paths to the Win32 format in some cases.

**Reply by stdedos** [reply]
Yep, that's it:
![image](https://user-images.githubusercontent.com/133706/109964384-4799e880-7cf6-11eb-8c09-6071f1607b80.png)

Awesome, thank you!

**Reply by stdedos** [reply]
PS: It doesn't solve the problem (still cannot create an image of the drive), but at least I know know, from PH's point of view, there are no `ImDisk` open handles.

Why does it complain and says it has no access idk (I tried UAC elevation, it does _something_, but it still complains too much / doesn't work afterwards)

**Comment by sskras** [comment]
@diversenok commented [on Mar 4, 2021](https://github.com/winsiderss/systeminformer/discussions/819#discussioncomment-429003):

> I remember [some compatibility issues](https://wj32.org/processhacker/forums/viewtopic.php?f=33&t=3419&p=10598#p10598) were preventing Process Hacker from correctly converting native paths to the Win32 format in some cases.

I am interested in the issue. Sadly the original site is unreachable to me, and the Internet Archive seems to [have no copy](https://web.archive.org/web/20201001135615/https://wj32.org/processhacker/forums/viewtopic.php?f=33&t=3419&sid=35640f25cebd5d498086930d10695415) of this particular page:

[Small bug in the autoupdate code of the beta channel?](https://web.archive.org/web/20201001135615/https://wj32.org/processhacker/forums/viewforum.php?f=33#3419:~:text=Small%20bug%20in%20the%20autoupdate%20code%20of%20the%20beta%20channel%3F)

![image](https://github.com/winsiderss/systeminformer/assets/7887758/01085901-d47a-449d-b77f-370c0dd4c4ac)
